File: backend/app/api/notifications.py
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, Query
from sqlalchemy import func, select, update
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.api.deps import get_current_user
from app.core.database import get_db
from app.models import Notification, User
from app.schemas import NotificationListOut, NotificationOut

router = APIRouter(prefix="/api/notifications", tags=["notifications"])


@router.get("", response_model=NotificationListOut)
async def list_notifications(
    page: int = Query(default=1, ge=1),
    size: int = Query(default=20, ge=1, le=100),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    base_query = select(Notification).where(Notification.user_id == current_user.id)
    total = (await db.execute(select(func.count()).select_from(base_query.subquery()))).scalar() or 0
    unread_count = (await db.execute(
        select(func.count()).where(
            Notification.user_id == current_user.id,
            Notification.is_read == False,
        )
    )).scalar() or 0

    result = await db.execute(
        base_query
        .options(selectinload(Notification.actor))
        .order_by(Notification.created_at.desc())
        .offset((page - 1) * size)
        .limit(size)
    )
    notifications = result.scalars().all()

    return NotificationListOut(
        items=[_to_notification_out(item) for item in notifications],
        total=total,
        unread_count=unread_count,
        page=page,
        page_size=size,
    )


@router.get("/unread-count")
async def unread_count(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    count = (await db.execute(
        select(func.count()).where(
            Notification.user_id == current_user.id,
            Notification.is_read == False,
        )
    )).scalar() or 0
    return {"unread_count": count}


@router.post("/mark-read")
async def mark_notifications_read(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await db.execute(
        update(Notification)
        .where(Notification.user_id == current_user.id, Notification.is_read == False)
        .values(is_read=True, read_at=datetime.now(timezone.utc))
    )
    await db.commit()
    return {"detail": "Marked as read"}


def _to_notification_out(notification: Notification) -> dict:
    actor = None
    if notification.actor:
        actor = {
            "id": notification.actor.id,
            "username": notification.actor.username,
            "email": notification.actor.email,
            "avatar_url": notification.actor.avatar_url,
            "bio": notification.actor.bio,
            "is_admin": notification.actor.is_admin,
            "created_at": notification.actor.created_at.isoformat() if notification.actor.created_at else "",
        }
    return {
        "id": notification.id,
        "user_id": notification.user_id,
        "actor_id": notification.actor_id,
        "type": notification.type,
        "blueprint_id": notification.blueprint_id,
        "comment_id": notification.comment_id,
        "payload": notification.payload,
        "is_read": notification.is_read,
        "created_at": notification.created_at.isoformat() if notification.created_at else "",
        "read_at": notification.read_at.isoformat() if notification.read_at else None,
        "actor": actor,
    }

File: backend-go/internal/handler/notifications.go
package handler

import (
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"

	"brickplans/internal/auth"
	"brickplans/internal/config"
	"brickplans/internal/db"
	"brickplans/internal/dto"
)

type NotificationsHandler struct {
	cfg *config.Config
	gdb *gorm.DB
}

func NewNotificationsHandler(cfg *config.Config, gdb *gorm.DB) *NotificationsHandler {
	return &NotificationsHandler{cfg: cfg, gdb: gdb}
}

func (h *NotificationsHandler) RegisterRoutes(rg *gin.RouterGroup) {
	g := rg.Group("/notifications", auth.AuthRequired(h.cfg, h.gdb))
	g.GET("", h.list)
	g.GET("/unread-count", h.unreadCount)
	g.POST("/mark-read", h.markRead)
}

func (h *NotificationsHandler) list(c *gin.Context) {
	user := auth.CurrentUser(c)
	page := atoiOr(c.Query("page"), 1, 1)
	size := atoiOr(c.Query("size"), 20, 1)
	if size > 100 {
		size = 100
	}

	var total int64
	h.gdb.Model(&db.Notification{}).Where("user_id = ?", user.ID).Count(&total)
	var unread int64
	h.gdb.Model(&db.Notification{}).Where("user_id = ? AND is_read = ?", user.ID, false).Count(&unread)

	var items []db.Notification
	h.gdb.Preload("Actor").Where("user_id = ?", user.ID).
		Order("created_at DESC").Offset((page - 1) * size).Limit(size).Find(&items)

	out := make([]dto.NotificationOut, 0, len(items))
	for i := range items {
		out = append(out, *toNotificationOut(&items[i]))
	}
	c.JSON(http.StatusOK, dto.NotificationListOut{
		Items:       out,
		Total:       int(total),
		UnreadCount: int(unread),
		Page:        page,
		PageSize:    size,
	})
}

func (h *NotificationsHandler) unreadCount(c *gin.Context) {
	user := auth.CurrentUser(c)
	var count int64
	h.gdb.Model(&db.Notification{}).Where("user_id = ? AND is_read = ?", user.ID, false).Count(&count)
	c.JSON(http.StatusOK, gin.H{"unread_count": count})
}

func (h *NotificationsHandler) markRead(c *gin.Context) {
	user := auth.CurrentUser(c)
	now := time.Now().UTC()
	h.gdb.Model(&db.Notification{}).
		Where("user_id = ? AND is_read = ?", user.ID, false).
		Updates(map[string]interface{}{"is_read": true, "read_at": now})
	c.JSON(http.StatusOK, gin.H{"detail": "Marked as read"})
}

func toNotificationOut(n *db.Notification) *dto.NotificationOut {
	var readAt *string
	if n.ReadAt != nil {
		s := dto.ISO(*n.ReadAt)
		readAt = &s
	}
	return &dto.NotificationOut{
		ID:          n.ID,
		UserID:      n.UserID,
		ActorID:     n.ActorID,
		Type:        n.Type,
		BlueprintID: n.BlueprintID,
		CommentID:   n.CommentID,
		Payload:     n.Payload,
		IsRead:      n.IsRead,
		CreatedAt:   dto.ISO(n.CreatedAt),
		ReadAt:      readAt,
		Actor:       dto.FromUser(n.Actor),
	}
}
